`UnitControl`: tidy up utils and types

ciampo opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

_While working on #38753 (specifically on the type of the `value` prop of the `NumberControl` component), it became clear that we should first untangle a few components using `NumberControl` — and `UnitControl` is one of them._

This PR tries to refactor and tidy up some of the types (and consequently the logic) of the `UnitControl`, in particular its many utility functions. **With that in mind, this PR's main goal is to add a clear internal split between the numeric quantity and the unit parts that together compose the value held by `UnitControl`**, all while introducing the least amount of runtime change possible and ideally no changes to the public API of `UnitControl` (despite it being flagged as an experimental component).

### Changes overview

- in order to avoid confusion around the meaning of "value", this PR renamed (and adjusted types) for a few variables/functions:
    - the value passed to `UnitControl`, which internally goes through the component's parsing logic, is now referred to as _"Raw Value"_
    - the numeric value parsed by the component's internal logic is now referred to as _"quantity",_ and is always assumed to be a `number` (when defined) 
    - the unit value parsed by the component's internal logic is now referred to as _"unit"_, and is always assumed to be a `string` (when defined) 
    - to recap: a _"raw value"_  passed to `UnitControl` (e.g `14`, `'2.5rem'`, `'600%'`...) is parsed by the internal logic and split into a _"quantity"_ (of type `number`) and a _"unit"_ (of type `string`)
    - some utility functions have also been renamed to be more explicit (and often more verbose), and their JS docs have been updated. This will hopefully add more clarity to what each utility function is supposed to do.
 - the `default` property of the `WPUnitControlUnit` has changed to be a `number`, following the type of a parsed  "quantity" (it was also already documented as such in the README)
- the type alias `Value` has been removed, and replaced by more explicit `string` and/or `number` types where necessary
- the type alias `WPUnitControlUnitList` has been removed, and has been replaced by `WPUnitControlUnit[]` (which means that those values can only be array, and can't be `false` anymore)
- the `value` prop of `UnitControl` was previously typed as required, this PR changes it to be non-required (it was already being omitted or passed as `null` in the repo, and advertised as such in the README anyway)
- the native `UnitControl` behaviour has slightly changed around how a value of `[]` for the `units` prop is handled (see [code changes](https://github.com/WordPress/gutenberg/pull/38987/files#diff-5eefbe2941ac9aaafe134fccc95d9b1a348058caee56abc658967e636c3f44ed)) 

### List of introduced runtime changes

- As a consequence of the simplification and removal of the `WPUnitControlUnitList` type, when passing a list of units,  all `false` values were replaced by `[]`
- When passing values to `UnitControl`, changed values from `null` to `undefined`
- Since the parsed _"quantity"_ is considered a `number`, the parsing logic internal to `UnitControl` now returns `undefined` if a quantity can't be parsed (it previously returned `''`). As a consequence, the fallback to `''` has been moved to each consumer the parsing utility (if necessary).
- The `default` value of all "base" units has changed from `''` to `undefined`
- As a consequence of the fact that `value` can now be `undefined` (and that it is already passed as `null` or `undefined` anyway across the repo), a few more checks and fallbacks have been added to the `UnitControl` component and some of its utility functions

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

- Project builds
- All tests pass
- Storybook stories work as expected
- Components using `UnitControl` work as expected, both in Storybook and in the editor:
  - `FontSizePicker`
  - `BoxControl`
  - `FocalPointPicker`
  - `ToolsPanel`
- Components using `UnitControl` in the `@wordpress/block-editor` package work as expected:
  - `BorderRadiusControl`
  - `LetterSpacingControl`
  - `LineHeightControl`
  - `BorderWidthEdit`
  - `GapEdit`
  - "Flow" layout inspector controls
  - Column block width control
  - Cover block dimension controls
  - Post featured Image's dimension controls
  - Search block dimensions
  - Spacer block dimensions controls
- Components using `UnitControl` in the `@wordpress/edit-site` package work as expected:
  - Controls in the border panel and dimensions panel

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

Refactor

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- N/A I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- N/A I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

ciampo committed:
UnitControl: tidy up `utils` types and logic

ciampo committed:
UnitControl: update components following the changes in the utils

ciampo committed:
UnitControl: Update tests

ciampo committed:
Update usages in `@wordpress/block-editor` package

ciampo committed:
Allow `rawValue` to be `undefined` (as originally) in `parseQuantityAndUnitFromRawValue`

ciampo committed:
Update usage of utils across `@wordpress/components` package

ciampo committed:
Use empty array instead of `false` for WPUnitControlUnit[] type

ciampo committed:
Default value of `` in `BoxControl`

ciampo committed:
`BoxControl`: use `undeefined` instead of `null` when sending values to `UnitControl`

ciampo committed:
`UnitCotnrol`: rewrite onChange handlers to be closer to original code, and fallback to `` more often

ciampo committed:
Add null check to parsing logic for added resiliance

ciampo committed:
Rename variable in border radius control

ciampo committed:
Rename variable in box control

ciampo committed:
Fallback parsed quantity to `` in UnitControl

ciampo commented on packages/components/src/box-control/utils.js:
```diff
@@ -25,10 +25,10 @@ export const LABELS = {
-	right: null,
-	bottom: null,
-	left: null,
+	top: undefined,
+	right: undefined,
+	bottom: undefined,
+	left: undefined,
```
Since these values are passed to `UnitControl`'s `value`, `undefined` is a more correct value here

ciampo commented on packages/components/src/box-control/utils.js:
```diff
@@ -85,7 +89,7 @@ export function getAllValue( values = {}, availableSides = ALL_SIDES ) {
 	 * simple truthy check.
 	 */
-	const allValue = isNumber( value ) ? `${ value }${ unit }` : null;
+	const allValue = isNumber( value ) ? `${ value }${ unit }` : undefined;
```
Same here re. passing `undefined` instead of `null` to `UnitControl`

ciampo committed:
Use `undefined` instead of `null` when passing values to `UnitControl` in `BorderRadiusControl`

ciampo committed:
`UnitControl`: allow `value` to be `undefined`, and add further resiliance agaist `null` values

ciampo committed:
Adjust border-radius-control unit tests to return `undefined` instead of `null`

ciampo committed:
Use empty array instead of `false` for list of units

ciampo committed:
Update README

ciampo committed:
CHANGELOG

ciampo commented on packages/block-editor/src/components/border-radius-control/test/utils.js:
```diff
@@ -121,26 +121,26 @@ describe( 'getAllValue', () => {
 		it( 'should return px when passed an array', () => {
-			expect( getAllValue( [] ) ).toBe( null );
+			expect( getAllValue( [] ) ).toBe( undefined );
 		} );
 		it( 'should return px when passed a boolean', () => {
-			expect( getAllValue( false ) ).toBe( null );
+			expect( getAllValue( false ) ).toBe( undefined );
 		} );
 		it( 'should return px when passed undefined', () => {
-			expect( getAllValue( false ) ).toBe( null );
+			expect( getAllValue( false ) ).toBe( undefined );
 		} );
```
I should probably push a couple of updates to these tests:

- the test names seem wrong (probably copy/paste)
- the last test should be probably changed to use `undefined` instead of `false`

ciampo replied:
Done

mirka commented on packages/block-editor/src/components/line-height-control/index.native.js:
```diff
@@ -2,6 +2,7 @@
  * WordPress dependencies
  */
 import { __ } from '@wordpress/i18n';
+// Is this even used? Shouldn't this named import be `__experimentalUnitControl` ?
```
TIL React Native has an [entirely separate index](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/index.native.js#L56) 😮

mirka commented on packages/block-editor/src/components/line-height-control/index.native.js:
```diff
@@ -20,7 +21,7 @@ export default function LineHeightControl( { value: lineHeight, onChange } ) {
 			value={ value }
 			onChange={ onChange }
-			units={ false }
+			units={ [] }
```
Based on the [separate index](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/index.native.js#L56), I'd assume this is the RN version of `UnitControl`, so we probably _shouldn't_ make this change? Like, I see some [runtime changes](https://github.com/WordPress/gutenberg/blob/08625a7be45d72c37bd4cc39a00a9ac68577bf73/packages/components/src/unit-control/index.native.js#L119-L148) that would happen.

ciampo replied:
You definitely make a good point, although this may lead to  more issues, since the native `UnitControl` component [shares the same utility functions](https://github.com/WordPress/gutenberg/blob/08625a7be45d72c37bd4cc39a00a9ac68577bf73/packages/components/src/unit-control/index.native.js#L19) of the web component.

Also, I think it would be overall be better to have the web and mobile counterpart share the same public APIs, since they also share the same README etc.

For now, I've gone ahead and removed all changes related to native components. That, in combination with the "more robust" checks that I've added back to the shared utils, should be good enough to avoid introducing regressions to the native components.

Having said that, I'm still not convinced that this is best solution. Ideally, we should update the native `UnitControl`'s behaviour not to rely on `units` being `false` at all — although I'm not in a position to say what's the best solution here, since it looks like the native `UnitControl` behaves differently when `false` is passed instead of `[]`. Tagging in @enejb and @lukewalczak who recently worked on this component for guidance.

mirka commented on packages/components/src/unit-control/utils.ts:
```diff
@@ -145,131 +145,137 @@ export const DEFAULT_UNIT = allUnits.px;
+	allowedUnits: WPUnitControlUnit[] = ALL_CSS_UNITS
+): [ number | undefined, string | undefined ] {
+	let trimmedValue;
+	let valueToReturn;
```
`valueToReturn` → `quantityToReturn`?

ciampo replied:
Yup makes sense! Done :)

mirka commented on packages/components/src/unit-control/utils.ts:
```diff
@@ -145,131 +145,137 @@ export const DEFAULT_UNIT = allUnits.px;
-		baseValue = fallbackValue;
-	}
+	// The parsed value from `parseQuantityAndUnitFromRawValue` should now be
+	// either a real number or undefined. If not, use the fallback value.
```
```suggestion
	// either a real number or undefined. If undefined, use the fallback value.
```

ciampo replied:
Reads better, thank you!

mirka commented on packages/components/src/unit-control/utils.ts:
```diff
@@ -282,32 +288,31 @@ export function filterUnitsWithSettings(
  * @param  args.defaultValues  Collection of default values for defined units. Example: { px: '350', em: '15' }.
  *
- * @return Filtered units based on settings.
+ * @return Filtered units of units, with their default values updated following the `defaultValues` argument's property.
```
```suggestion
 * @return Filtered list of units, with their default values updated following the `defaultValues` argument's property.
```

...I think is what we mean?

ciampo replied:
Haha yep! I'm not even sure what "units of units" may be

mirka commented on packages/components/src/unit-control/types.ts:
```diff
@@ -28,7 +26,7 @@ export type WPUnitControlUnit = {
 	/**
 	 * Default value for the unit, used when switching units.
```
This confused me for a bit, 👇  might've been easier to understand at first glance:

```suggestion
	 * Default value (quantity) for the unit, used when switching units.
```

ciampo replied:
Makes sense, the use of the word "value" in this component is tricky and confusing

mirka reviewed (commented):
I love these improvements, it's like a masterclass in using empty arrays and undefined instead of false/null to simplify conditionals 💯 And as someone who's reading this code for the first time, the verbosity helps a lot. Beautiful work!

I mainly approached this as a code reader, so although it looks like we have decent test coverage around the runtime changes, I'll defer to others who've worked with `UnitControl` more intimately to take a better look at the behavioral aspects.

ramonjd commented:
Things are working well for me. I tested changing units, changing individual box control values. All looks as it should in the frontend and editors.

I've tested the following:

#### Components using UnitControl in the @wordpress/block-editor package work as expected:

✅  BorderRadiusControl
✅  LetterSpacingControl
✅  LineHeightControl
✅  BorderWidthEdit
✅  GapEdit
✅  Column block width control
✅  Cover block dimension controls
✅  Spacer block dimensions controls
✅ Search block dimensions
✅ Post featured Image's dimension controls

#### Components using UnitControl in the @wordpress/edit-site package work as expected:

✅ Controls in the border panel and dimensions panel

- FocalPointPicker
- Padding
- Margin
- Block spacing
- Custom font size
- Border width and radius

andrewserong commented on packages/components/src/unit-control/index.tsx:
```diff
@@ -75,11 +85,15 @@ function UnitControl(
-		if ( next === '' ) {
+		if (
+			nextQuantityValue === '' ||
+			typeof nextQuantityValue === 'undefined' ||
```
Just out of curiosity, why is this a `typeof` and checking against the string value `'undefined'` instead of `===` the value `undefined`? (I wasn't sure if this is just a style preference or if there was another reason 🙂)

ciampo replied:
Good question!

It's a personal preference, a sort of "safe default" to be more on the side of caution, since the `typeof` operator works with undeclared variables, while the identical operator will throw a `ReferenceError` exception.

```js
typeof undeclaredVar === 'undefined'; // true
undeclaredVar === undefined; // throws a ReferenceError exception
```

In this case, it won't make a difference, since `nextQuantityValue` is initialised as an argument of `handleOnQuantityChange`. Just a habit I picked up some time ago!

andrewserong replied:
Ah, perfect (TIL), thanks for the explanation, I'm glad I asked!

andrewserong commented on packages/components/src/unit-control/test/utils.js:
```diff
@@ -65,27 +65,30 @@ describe( 'UnitControl utils', () => {
 		} );
 
-		it( 'should return empty array where available units is set to false', () => {
+		it( 'should return empty array where available units is set to an empty array', () => {
```
Would it be worth keeping this test around as-is, and also including an additional test that passes in `availableUnits` as an empty array?

Previously the function gracefully handled any value that wasn't an array (and returned an empty array), so I wasn't sure if it's worth still covering that case.

ciampo replied:
Good point, I kept the old test as-is and added a new test for passing an empty array

andrewserong commented on packages/components/src/unit-control/types.ts:
```diff
@@ -12,8 +12,6 @@ import type {
 	Size as InputSize,
 } from '../input-control/types';
 
-export type Value = number | string;
```
Lovely, nice to be able to remove this one! 😍

ciampo replied:
I know right! This also goes in the direction discussed in #38842 around the fact that the default docgen used by Storybook to extract TS props into docs can not expand type aliases

andrewserong commented on packages/components/src/unit-control/utils.ts:
```diff
@@ -145,131 +145,137 @@ export const DEFAULT_UNIT = allUnits.px;
+ * @return Whether the list actually contains any units.
  */
-export function hasUnits( units: WPUnitControlUnitList ): boolean {
-	return Array.isArray( units ) && !! units.length;
```
Nit: Should we keep the `Array.isArray()` check here? This is quite a contrived example, but otherwise if the function is called with a string, it will return `true`.

Edit: given that we're using TypeScript and the `hasUnits` function isn't being exported outside of the package, this is probably fine to ignore! 😅

ciampo replied:
Makes sense, you raise a good point! I've added the check back

andrewserong commented on packages/components/src/unit-control/utils.ts:
```diff
@@ -145,131 +145,137 @@ export const DEFAULT_UNIT = allUnits.px;
+ * @param  allowedUnits     Units to derive from.
+ * @param  fallbackQuantity The fallback quantity, used in case it's not possible to parse a valid quantity from the raw value.
+ * @param  fallbackUnit     The fallback unit, used in case it's not possible to parse a valid unit from the raw value.
+ * @return The extracted quantity and unit. The quantity can be `undefined` in case the raw value could not be parsed to a number correctly, and the `fallbackQuantity` was also `undefined`. The unit can be `undefined` only if the unit parsed from the raw value could not be matched against the list of allowed units, the `fallbackQuantity` is also `undefined` and the list of `allowedUnits` is passe ...
```
Great explanation here in the return statement, it'll make it much easier for folks reading this function in the future!

Tiny nit: is it worth adding some formatting here to span multiple lines for views without word-wrap?

ciampo replied:
Sure thing, I've broken longer JSDoc comments over multiple lines in this case a few more across this file.

andrewserong commented on packages/components/src/unit-control/utils.ts:
```diff
@@ -145,131 +145,137 @@ export const DEFAULT_UNIT = allUnits.px;
-	unitSetting: Array< string > = [],
-	units: WPUnitControlUnitList
-): Array< WPUnitControlUnit > {
-	return Array.isArray( units )
```
Should we keep any of the defensive checks here in case we receive a value that isn't an array? `useCustomUnits` is called from a bunch of places, and we currently expect that callers will ensure that an array value is always passed (e.g. [here in the margin support](https://github.com/wordpress/gutenberg/blob/23369ee14e567eb32a8a004f6dd1c3d867a863e5/packages/block-editor/src/hooks/margin.js#L97)), but I wasn't sure if this is placing too much onus upon the calling code, and whether we need to gracefully handle illegal values?

ciampo replied:
I agree with your reasoning here, better to stay on the side of caution. This basically also avoids introducing a runtime change, and still keeps TypeScript happy :)

andrewserong commented on packages/components/src/unit-control/utils.ts:
```diff
@@ -282,32 +288,31 @@ export function filterUnitsWithSettings(
-		! availableUnits ? [] : availableUnits,
+	units?: WPUnitControlUnit[];
+	availableUnits?: string[];
+	defaultValues: Record< string, number >;
```
It makes sense that this type should be `number`, but I see in a few places in the codebase there are instances where it's passed in a string value, [such as here](https://github.com/wordpress/gutenberg/blob/922976bbe95c4ab4da06cb9d7e5e25d98f1d398c/packages/block-library/src/cover/edit.js#L113). Should we allow that, or are we happy for those usages to technically be errors? The value will ultimately be coerced to a string, so there might not be much harm in using the following?

```suggestion
	defaultValues: Record< string, string | number >;
```

ciampo replied:
I understand your suggestion, but I think it's a bit tricky since this type is directly tied to the fact that we expect a `quantity` to be of type `number`.

I've taken an alternative approach and:
- refactored all of the `useCustomUnits` instances where `string` where still used instead of `number` (which I had missed until your comment, thank you!)
- I've added a unit test that checks for string values, for added resiliency

What do you think?

andrewserong replied:
The looks great to me! Thanks for adding in the unit test and updating those other usages — this gives us the desired behaviour of a consistent API ("this is how the component should be used") while still preserving accepting string values, so it's a nice balance 👍

andrewserong commented on packages/components/src/unit-control/utils.ts:
```diff
@@ -282,32 +288,31 @@ export function filterUnitsWithSettings(
+	defaultValues: Record< string, number >;
+} ): WPUnitControlUnit[] => {
+	const customUnitsToReturn = filterUnitsWithSettings(
+		availableUnits,
```
We're no longer checking in `filterUnitsWithSettings` that `availableUnits` is an array, so could this change which passes the value in directly without the fallback `[]` be potentially risky if someone accidentally passes a nullish value as `availableUnits`?

ciampo replied:
Good point! The changes to `filterUnitsWithSettings` (adding the `isArray` check back) prompted by your previous comment should cover this scenario if I'm correct!

andrewserong commented on packages/components/src/unit-control/utils.ts:
```diff
@@ -318,36 +323,32 @@ export const useCustomUnits = ( {
-	legacyUnit: string | undefined,
-	units: WPUnitControlUnitList = ALL_CSS_UNITS
-): WPUnitControlUnitList {
-	if ( ! Array.isArray( units ) ) {
```
This function now also expects that `UnitControl` will never be passed a nullish value for its list of units. Is that safe?

ciampo replied:
I've added back the `Array.isArray` check, although I haven't reintroduced the early return  statement — this means that if `units` is null or not an array, the function will still return an array of units (empty in the worst case scenario).

Do you think it's safe enough?

andrewserong replied:
Nice, yes that gives us a situation where we're gracefully failing to a desired value. LGTM!

andrewserong reviewed (commented):
What a great PR, thanks for all your work here tidying up the component, I think this greatly improves the readability of the component and its utility functions! ✨

I haven't been able to find any actual issues with using the component in practice, but I've added some comments about where there are some runtime changes surrounding the expected value for `units` and the removal of some `Array.isArray()` checks. Most of the feedback is totally optional, of course, and sort of point more to the question of how careful should the component be about dealing with potential values that it's passed? Because it's such a heavily used component, I think I lean a bit toward a cautious approach of leaving some of those checks in, but I do appreciate it might be cleaner to remove some of them, too!

Also, just to echo Lena's feedback, I love the shift to `undefined` in most of the areas, and using it as a valid value (without accidentally rendering it as a string) — I think that fix probably explains why `null` might have been used in lots of places in the past?

The careful function renaming and added inline comments are really great, too, gives the component a much-needed polish. Kudos!

ntsekouras commented on packages/components/src/unit-control/utils.ts:
```diff
@@ -145,131 +145,137 @@ export const DEFAULT_UNIT = allUnits.px;
  */
-export function hasUnits( units: WPUnitControlUnitList ): boolean {
-	return Array.isArray( units ) && !! units.length;
+export function hasUnits( units?: WPUnitControlUnit[] ): boolean {
```
Do we really need this util? It's used five times from which three of them also check the `length` again.

ciampo replied:
I was trying to introduce the least amount of runtime changes possible — and so I'd prefer to keep this function around.

I've also added back the `isArray` check as per Andrew's comment which makes this function a bit more compelling, I think!

ntsekouras commented on packages/components/src/unit-control/utils.ts:
```diff
@@ -145,131 +145,137 @@ export const DEFAULT_UNIT = allUnits.px;
-	let unit: string | undefined =
-		unitMatch?.[ 1 ] !== undefined ? unitMatch[ 1 ] : '';
-	unit = unit.toLowerCase();
+	if ( typeof rawValue !== 'undefined' || rawValue === null ) {
```
Similar comment with @andrewserong about `typeof` check.

ciampo replied:
I left a reply to Andrew's comment — TL;DR: personal preference/habit to stay more on the side of caution

ntsekouras commented on packages/components/src/unit-control/utils.ts:
```diff
@@ -145,131 +145,137 @@ export const DEFAULT_UNIT = allUnits.px;
+		trimmedValue = `${ rawValue }`.trim();
+		const parsedQuantity = parseFloat( trimmedValue );
+		valueToReturn =
+			isNaN( parsedQuantity ) || ! isFinite( parsedQuantity )
```
Why both checks? `NaN` is handled by `isFinite`.

ciampo replied:
Good point — I removed the `NaN` check and only left `isFinite`

ntsekouras commented on packages/components/src/unit-control/utils.ts:
```diff
@@ -145,131 +145,137 @@ export const DEFAULT_UNIT = allUnits.px;
-		const match = units.find( ( item ) => item.value === unit );
-		unit = match?.value;
+	const unitMatch = trimmedValue?.match( /[\d.\-\+]*\s*(.*)/ );
+	const matchedUnit =
```
We can simplify with optional chaining here

ciampo replied:
Nice, I didn't notice it while refactoring the code. Much more readable!

ntsekouras commented on packages/components/src/unit-control/index.tsx:
```diff
@@ -88,53 +102,64 @@ function UnitControl(
 		);
 
-		refParsedValue.current = parsedValue.toString();
+		refParsedValue.current = ( validParsedQuantity ?? '' ).toString();
```
```suggestion
		refValidParsedQuantity.current = ( validParsedQuantity ?? '' ).toString();
```

ciampo replied:
Well spotted!

This actually prompted me to do a slightly bigger refactor:

- renamed `refParsedValue` to `refParsedQuantity`
- changed the ref type from being `string | null` to `number | undefined` (to match the type of `quantity` more closely)
- the conversion from `number` to `string` only happens when committing that ref value to the unit control state reducer

Much cleaner!

ntsekouras commented on packages/components/src/unit-control/utils.ts:
```diff
@@ -7,113 +7,113 @@ import { Platform } from '@wordpress/element';
 		value: 'px',
 		label: isWeb ? 'px' : __( 'Pixels (px)' ),
-		default: '',
+		default: undefined,
```
Do we need to explicitly set these to `undefined`?

ciampo replied:
No, we don't — I removed all instances from the code.

ntsekouras reviewed (commented):
Thanks for the PR Marco! We always need some simplifications and that make the code more maintainable, kudos! 🚀 

I've tested and everything seems to be working well! All my comments are super small nit picks.

ciampo committed:
Border control: update test names

ciampo committed:
Native components: remove unnecessary comments, fix default value of units

ciampo committed:
Undo native components runtime change

ciampo committed:
`valueToReturn` => `quantityToReturn`

ciampo committed:
Clarify comment

ciampo committed:
No more units of units

ciampo committed:
Clarify default unit docs

ciampo committed:
Reinstantiate the `filterUnitsWithSettings` test chacking for `false` inputs

ciampo committed:
Break long JSDocs over multiple lines

ciampo committed:
Update `useCustomUnits` usages to pass numbers for the default unit quantity values

ciampo committed:
Reintroduce `isArray` check in `getUnitsWithCurrentUnit`

ciampo committed:
Remove unnecessary `NaN` check, redundant next to `isFinite`

ciampo committed:
Simply `matchedUnit` computation using optional chaining syntax

ciampo committed:
Remove `undefined` default unit quantity values

ciampo reviewed (commented):
Thank you all! I'm humbled by all of your thoughtful and detailed reviews 🙇 

I've pushed changes and replied to all comments, trying to be as granular as possible with my commits to facilitate reviewing.

> Most of the feedback is totally optional, of course, and sort of point more to the question of how careful should the component be about dealing with potential values that it's passed? Because it's such a heavily used component, I think I lean a bit toward a cautious approach of leaving some of those checks in, but I do appreciate it might be cleaner to remove some of them, too!

You definitely make a good point, and I agree with you on this one. Using TypeScript _internally_ to `UnitControl` doesn't necessarily prevent the component from being used incorrectly in non-typed environments. And since there may already be a few instances out there, it's probably best to stay on the side of caution and re-introduce a few more checks.

andrewserong commented:
Nice work making all those updates @ciampo, I've replied and resolved all the feedback I gave (looks like you've addressed it all), and this PR is looking in great shape to me. I haven't had a chance to re-test it in detail yet, but happy to do that tomorrow. Also, looks like it needs a rebase now due to other changes in `unit-control/index.tsx`.

Thanks again for all the detailed work here!

ciampo committed:
Add prop types reminder

ciampo commented:
> Thanks again for all the detailed work here!

My pleasure! There's going to be more of this in the upcoming weeks, with the ongoing TypeScript migration, so I really appreciate your help here!

> [...] this PR is looking in great shape to me. I haven't had a chance to re-test it in detail yet, but happy to do that tomorrow. Also, looks like it needs a rebase now

Thank you!  I've rebased and solved the conflict.

Here's what's left before merging this PR:

- a final round of smoke tests for the "web" part
- the issue raised in [this comment](https://github.com/WordPress/gutenberg/pull/38987#discussion_r814684970) on the "native" side

enejb commented:
Thanks for working on this @ciampo. 

I agree that the mobile native side should follow the same conventions as the web (even if the UI is different) but we have to be a bit careful that we don't break things. Please do make the changes so that things match up. 

Have you tried to run your code in the example native app that comes with Gutenberg? Here is a guide that helps you get started https://developer.wordpress.org/block-editor/contributors/code/react-native/osx-setup-guide/ ?   

cc: @hypest for awareness.

ciampo commented:
> Thanks for working on this @ciampo.
> 
> I agree that the mobile native side should follow the same conventions as the web (even if the UI is different) but we have to be a bit careful that we don't break things. Please do make the changes so that things match up.
> 
> Have you tried to run your code in the example native app that comes with Gutenberg? Here is a guide that helps you get started https://developer.wordpress.org/block-editor/contributors/code/react-native/osx-setup-guide/ ?
> 
> cc: @hypest for awareness.

Hey @enejb ! Sure, I can go ahead and refactor the code — I'd just need some guidance on this aspect:

> Ideally, we should update the native `UnitControl`'s behaviour not to rely on units being `false` at all — although I'm not in a position to say what's the best solution here, since it looks like the native `UnitControl` behaves differently when `false` is passed instead of `[]`

In particular:
- [when `units` has a value of `false`, the `renderUnitPicker` function returns `null`](https://github.com/WordPress/gutenberg/blob/f90894082e81bbedcfb56441f1d9298f6d292667/packages/components/src/unit-control/index.native.js#L119-L121), while
- when `units` has a value of `[]`, [the `renderUnitPicker` function would currently returns a `View`, but wouldn't show the `<Picker />` component](https://github.com/WordPress/gutenberg/blob/f90894082e81bbedcfb56441f1d9298f6d292667/packages/components/src/unit-control/index.native.js#L122-L135)

My instinct here would be to change that `false` check, and have the `renderUnitPicker` return `null` also when `units` is an empty array. Does it make sense to you?

andrewserong reviewed (approved):
Looks great! This is testing well from the web side of things:

✅ Project builds
✅ All tests pass
✅ Storybook stories work as expected
✅ Components using UnitControl work as expected, both in Storybook and in the editor:
✅  - `FontSizePicker`
✅  - `BoxControl`
✅  - `FocalPointPicker`
✅  - When used within the `ToolsPanel`

✅ Components using UnitControl in the @wordpress/block-editor package work as expected:
✅  - `BorderRadiusControl`
✅  - `LetterSpacingControl`
✅  - `LineHeightControl`
✅  - `BorderWidthEdit`
✅  - `GapEdit`
✅ "Flow" layout inspector controls
✅ Column block width control
✅ Cover block dimension controls
✅ Post featured Image's dimension controls
✅ Search block dimensions
✅ Spacer block dimensions controls

✅ Components using UnitControl in the @wordpress/edit-site package work as expected:
✅ Controls in the border panel and dimensions panel
✅ Settings custom units (both empty, and a different set of units) in `settings.spacing.units` in `theme.json` works correctly

Once you've got confirmation from the native side of things, this looks good to merge to me! 🚀

ciampo committed:
Update react native files

ciampo commented:
Thank you @andrewserong , your in-depth review has been very helpful!

> Once you've got confirming from the native side of things, this looks good to merge to me! 🚀

I've spent some time setting my local environment to run the native editor, but I wasn't able to find how to reach the [native `LineHeightControl` component](https://github.com/WordPress/gutenberg/blob/24077bd27dd31610d687289d926a19055e62644d/packages/block-editor/src/components/line-height-control/index.native.js) (or even the [native `UnitControl` component](https://github.com/WordPress/gutenberg/blob/24077bd27dd31610d687289d926a19055e62644d/packages/components/src/unit-control/index.native.js)).

I still decided to go ahead and apply [the changes that I proposed yesterday to the native `UnitControl`](https://github.com/WordPress/gutenberg/pull/38987/commits/24077bd27dd31610d687289d926a19055e62644d). The main difference is that the `renderUnitPicker` function now returns `null` in case `units` is either `false` or `[]`. As a comparison, previously the `renderUnitPicker` function would instead still render `renderUnitButton` (but no picker) when the passed `units` are `[]`.

@enejb and @hypest , would you be so kind to review and test this last part in native land, and hopefully give the green light for this PR to get merged? Thank you!

hypest commented:
Thanks for the work and the ping @ciampo !

I have triggered the [automated PR](https://github.com/wordpress-mobile/gutenberg-mobile/pull/4618) with tests on the gutenberg-mobile repo side and I'll have a look at the changes as well! Update: CI is green so far on that PR and I just now triggered the extended UI tests job there. Update: All CI tests are green!

For the UnitControl I'd look at the Spacer block (and its settings for the space height). For the LineHeightControl, perhaps @geriux you can in the meantime share some code pointers and testing instructions? Thanks!

ciampo committed:
Rename variable after rebase

ciampo committed:
Rename function in tests after rebase

ciampo commented:
Update: I rebased this PR (and renamed a variable) to include changes from #39148.

Regarding the React Native runtime changes, I've decided to remove those changes from this PR as they were causing a regression (see details below). I will tackle those changes separately, so that they don't delay this PR further (cc @geriux @hypest @enejb — I'll ping you once the new PR is ready).

Since the rest of these changes have already been approved, I'll wait for CI to go green (both on this PR and on https://github.com/wordpress-mobile/gutenberg-mobile/pull/4618) and then merge.

geriux commented:
> Regarding the React Native runtime changes, I've decided to remove those changes from this PR as they were causing a regression (see details below). I will tackle those changes separately, so that they don't delay this PR further (cc @geriux @hypest @enejb — I'll ping you once the new PR is ready).

Thanks @ciampo!

hypest commented:
> Regarding the React Native runtime changes, I've decided to remove those changes from this PR as they were causing a regression (see details below)

Works for me @ciampo , thank you for that strategy! By the way, curious to ask, was that a regression you spotted by manual testing or some other means? Thanks!

ciampo commented:
Forgot to add it to my previous comment! Basically, my changes were unintentionally hiding the part of the UI showing the unit (including the unit picker). Here's an example for the `Spacer` block:

| `trunk` | regression |
|---|---|
| ![Screenshot 2022-03-02 at 13 08 32](https://user-images.githubusercontent.com/1083581/156371650-95ea37a4-be80-41fc-9329-d893e86f3f6e.png) | ![Screenshot 2022-03-02 at 13 28 04](https://user-images.githubusercontent.com/1083581/156371809-e29ab40b-5101-49c1-89ff-c5e2e01bbb94.png) |
 
In that new PR, we will need to investigate better the logic around checking the `units` value. Thankfully, @geriux should be able to help with that :)
